plugins/com.google.cloud.tools.eclipse.appengine.login/src/com/google/cloud/tools/eclipse/appengine/login/AuthorizationCodeReceiver.java

```diff
@@ -0,0 +1,139 @@
+ * the License.
+ *******************************************************************************/
+
+package com.google.cloud.tools.eclipse.appengine.login;
+
+import com.google.common.base.Preconditions;
+
+import org.eclipse.jetty.server.Request;
+import org.eclipse.jetty.server.Server;
+import org.eclipse.jetty.server.ServerConnector;
+import org.eclipse.jetty.server.handler.AbstractHandler;
+
+import java.io.IOException;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+/**
+ * Runs a Jetty web server instance that opens up an HTTP listening port to receive OAuth2
+ * authorization code.
+ */
+public class AuthorizationCodeReceiver {
+
+  private static final Logger logger = Logger.getLogger(AuthorizationCodeReceiver.class.getName());
+
+  // TODO: create landing pages.
+  private static final String SUCCESS_LANDING_PAGE = "https://cloud.google.com/tools/intellij/auth_success";
+  private static final String FAILURE_LANDING_PAGE = "https://cloud.google.com/tools/intellij/auth_failure";
+
+  private static final String LOCAL_HOST = "127.0.0.1";
+  private static final String CODE_RECEPTION_POINT_PATH = "/code-reception-point";
+
+  private Server server;
+
+  private volatile boolean resultConsumed;
+
+  // Reads on these shared objects can happen only after server shutdown,
+  // so synchronization is not really necessary.
+  private String error;
+  private String authorizationCode;
+
+  /**
+   * Starts a local server listening through an ephemeral port.
+   *
+   * @return "http://127.0.0.1:<listening port>/code-reception-point", where this server listens
+   *     and expects to get an authorization code returning from a OAuth login flow
+   * @throws IOException wraps any {@link Exception} from Jetty when it failed to start
+   */
+  public String runLocalServer() throws IOException {
+    Preconditions.checkState(server == null, "Don't reuse the receiver.");
+
+    server = new Server(0);
+    server.setHandler(new RequestHandler());
+    ServerConnector connector = (ServerConnector) server.getConnectors()[0];
+    connector.setHost(LOCAL_HOST);
+    try {
+      server.setStopAtShutdown(true);
+      server.start();
+      logger.log(Level.INFO, "Started a local web server waiting for OAuth2 authorization code.");
+      return "http://127.0.0.1:" + connector.getLocalPort() + CODE_RECEPTION_POINT_PATH;
+    } catch (Exception ex) {
+      throw new IOException(ex);
+    }
+  }
+
+  /**
+   * Waits for a login result to arrive for the given timeout or until interrupted.
+   *
+   * @param timeout in millisecond. Cannot be 0
+   * @return true if the web server consumed the login result
+   */
+  public boolean waitForCode(long timeout) {
+    Preconditions.checkArgument(timeout != 0);
+    Preconditions.checkState(server != null, "Call runLocalServer() first.");
```

Comment:
Turns out that `waitForCode()` fits well as a polling method in an async progress thread that continuously monitor if the auth code has arrived.

`runLocalServer()` on the other hand is better called in the main thread. If there was an error running a server, we halt. Otherwise, it spins up the async thread that polls the server using `waitForCode()`.

So, I think separating these two methods seems better.